snow/engine/snowman/engine_decorator.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package snowman
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+)
+
+type DecoratedEngine struct {
+	*Engine
+	sd *stragglerDetector
+	f  func(time.Duration)
+}
+
+func NewDecoratedEngine(e *Engine, time func() time.Time, f func(time.Duration)) common.Engine {
```

Comment:
should we call this something else like `WithStragglerDetection` or `NewStragglerDetector`, sine Decorated seems more generic.
Also seems we could unexport the `decoratedEngine` struct, since we're returning `common.Engine`

Reply:
> It seems we do the calculation of requiring 0.8 of the stake both in getNetworkSnapshot and getNetworkInfo, could this computation be simplified somehow?

I can do it if you feel strongly about it, it's quite easy to do that.

However after thinking about it, I'm not sure that's ideal. I know the code can be made simpler by not repeating the computation, but, I structured it because I wanted the logs to distinguish between the following two cases:

- We do not know yet the last accepted block for enough nodes (stake) 
- We know the last accepted block for enough nodes (stake) but most of them is exactly like our own last accepted block.

I can easily remove the code handling the first case, but then if we don't proceed checking if we're behind the rest, the logs may falsely state it is because our last accepted block is as the rest, but it might be that we simply haven't gathered enough info.